sweep/sweeper.go

```diff
@@ -29,13 +64,387 @@ type UtxoSweeperConfig struct {
+	// transaction to the appropriate network.
+	PublishTransaction func(*wire.MsgTx) error
+
+	// Ticker is an external dependency that drives the batch window logic.
+	Ticker ticker.Ticker
+
+	// BatchWindowSec indicates the size of the batching time window in
+	// seconds.
+	BatchWindowSec int
+
+	// SweepTxConfTarget assigns a confirmation target for sweep txes on
+	// which the fee calculation will be based.
+	SweepTxConfTarget uint32
+
+	// Notifier is an instance of a chain notifier we'll use to watch for
+	// certain on-chain events.
+	Notifier chainntnfs.ChainNotifier
+
+	// ChainIO allows us to query the state of the current main chain.
+	ChainIO lnwallet.BlockChainIO
+
+	HasSpendingTx func(op wire.OutPoint) (*wire.MsgTx, error)
 }
 
-// New returns a new UtxoSweeper instance.
+// Result is the struct that is pushed through the result channel. Callers
+// can use this to be informed of the final sweep result. In case of a remote
+// spend, ErrRemoteSpend is return in err.
+type Result struct {
+	Err error
+	Tx  *wire.MsgTx
+}
+
+type sweeperInput struct {
+	input      Input
+	resultChan chan Result
+}
+
+type pendingInput struct {
+	txHash        chainhash.Hash
+	listeners     []chan Result
+	ntfnRegCancel func()
+}
+
+// New returns a new Sweeper instance.
 func New(cfg *UtxoSweeperConfig) *UtxoSweeper {
 	return &UtxoSweeper{
-		cfg: cfg,
+		cfg:           cfg,
+		newInputs:     make(chan *sweeperInput),
+		spendChan:     make(chan *chainntnfs.SpendDetail),
+		quit:          make(chan struct{}),
+		pendingInputs: make(map[wire.OutPoint]*pendingInput),
+	}
+}
+
+// Start starts the process of constructing and publish sweep txes.
+func (s *UtxoSweeper) Start() error {
+
+	if !atomic.CompareAndSwapUint32(&s.started, 0, 1) {
+		return nil
+	}
+
+	log.Debugf("Starting sweeper")
+
+	s.wg.Add(1)
+	go s.collector()
+
+	return nil
+}
+
+// Stop stops sweeper from listening to block epochs and constructing sweep
+// txes.
+func (s *UtxoSweeper) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	log.Infof("Sweeper shutting down")
+
+	close(s.quit)
+	s.wg.Wait()
+
+	log.Infof("Sweeper shut down")
+
+	return nil
+}
+
+// SweepInput sweeps inputs back into the wallet. The inputs will be batched and
+// swept after the batch time window ends.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	log.Debug("Pushing new input to collector")
+
+	sweeperInput := &sweeperInput{
+		input:      input,
+		resultChan: make(chan Result),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+func (s *UtxoSweeper) collector() {
+	log.Debug("Collector started")
+	defer s.wg.Done()
+
+	s.cfg.Ticker.Resume()
+	defer s.cfg.Ticker.Stop()
+
+	inputToSweep := make(map[wire.OutPoint]Input)
+	clearInputsToSweep := func() {
```

Comment:
Is this closure really necessary? Seems simple enough to just re-assign the slice...

Reply:
I was in doubt. Mainly added it to prevent people from mistakenly setting it to nil. But as you doubt it too, I removed it.